### PR TITLE
Allow custom transformations of assets

### PR DESCRIPTION
**Pull request by tim-smart**
This exposes one query parameter `transforms`, which is a JSON array of
sharp transformation methods + arguments.

It also updates the system asset presets.

### CODE REVIEW

File: api/src/services/assets.ts [hunk 1]
```diff
@@ -119,27 +154,12 @@ export class AssetsService {
...
-			'__' +
-			Object.entries(transformation)
-				.sort((a, b) => (a[0] > b[0] ? 1 : -1))
-				.map((e) => e.join('_'))
-				.join(',')
-		);
+		return [
+			'',
+			...transformation.map(([method, ...args]) => `${method}_${JSON.stringify(args).replace(/[^A-Za-z0-9]/g, '')}`),
+		].join('__');
```
Comment (tim-smart) [inline api/src/services/assets.ts:164]: Possible to normalize the options and create some sort of hash?

File: api/src/database/system-data/fields/settings.yaml [hunk 2]
```diff
@@ -125,62 +125,116 @@ fields:
...
+                        - value: removeAlpha
+                          text: removeAlpha
+                        - value: ensureAlpha
+                          text: ensureAlpha
+                        - value: extractChannel
+                          text: extractChannel
+                        - value: bandbool
+                          text: bandbool
+                  width: half
+                - field: options
```
Comment (tim-smart) [inline api/src/database/system-data/fields/settings.yaml:218]: There has to be a better way lol.

Comment (tim-smart) [reply api/src/database/system-data/fields/settings.yaml:290]: Partially kept old easy to use fields.

File: api/src/services/assets.ts (hunk 1 above)
Comment (tim-smart) [reply api/src/services/assets.ts]: Used `object-hash` from NPM instead.

**Comment by rijkvanzanten** [comment]
Love it! Thanks @tim-smart 🙂

**Comment by tim-smart** [comment]
I'm going to stop here on this one.

Not sure if you liked the direction this was taking?

We will need this in a week or two, so will float this patch on a fork until Directus has something for reformatting images etc.

**Comment by rijkvanzanten** [comment]
> Not sure if you liked the direction this was taking?

I think the direction of putting all the options under a single `transform` query param makes total sense. 


Haven't had the time to do a full review yet (I'm OOO currently), but this all looks very promising on first glance 🙂

**Comment by SeanDylanGoff** [comment]
@tim-smart key-value pairs instead of an array won't work because order matters for the transforms, right?
Because if we did have an object we could use the existing filter implementation for advanced configuration on which transforms are allowed and which aren't

**Comment by tim-smart** [comment]
> @tim-smart key-value pairs instead of an array won't work because order matters for the transforms, right?
> Because if we did have an object we could use the existing filter implementation for advanced configuration on which transforms are allowed and which aren't

Yes, order matters. You can also run the same transformation function more than once - you can't have duplicate keys in an object :)

File: api/src/controllers/assets.ts [hunk 3]
```diff
@@ -68,21 +68,54 @@ router.get(
...
-		if ('quality' in transformation && (Number(transformation.quality) < 1 || Number(transformation.quality) > 100)) {
-			throw new InvalidQueryException(`"quality" Parameter has to between 1 to 100`);
+
+		if ('transforms' in transformation) {
+			let transforms: unknown;
+
+			// Try parse the JSON array
+			try {
+				transforms = JSON.parse(transformation['transforms'] as string);
+			} catch (_) {
```
Comment (rijkvanzanten) [inline api/src/controllers/assets.ts:78]:
```suggestion
			} catch {
```

File: api/src/database/system-data/fields/settings.yaml [hunk 4]
```diff
@@ -181,6 +181,134 @@ fields:
...
+                        - value: toFormat
+                          text: toFormat
+                        - value: jpeg
+                          text: jpeg
+                        - value: png
+                          text: png
+                        - value: tiff
+                          text: tiff
+                        - value: webp
+                          text: webp
```
Comment (rijkvanzanten) [inline api/src/database/system-data/fields/settings.yaml:234]: These are already listed under `format` above, no need to duplicate here

File: api/src/database/system-data/fields/settings.yaml [hunk 5]
```diff
@@ -181,6 +181,134 @@ fields:
...
+                      template: '{{argument}}'
+                      fields:
+                        - field: argument
+                          name: Argument
+                          type: string
+                          schema:
+                            is_nullable: false
+                          meta:
+                            interface: input-code
+                            note: JSON representation of the argument
```
Comment (rijkvanzanten) [inline api/src/database/system-data/fields/settings.yaml:306]: We should figure out how to document this, little difficult to understand how to use it

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]
Little hard to understand how to use it exactly without having docs.

I'll take care of resolving the comments 👍🏻

**Comment by rijkvanzanten** [comment]
@tim-smart This is great, thanks!

The only thing I'm slightly worried about is the complexity of the parameter. Something like `transforms=[["resize", [200, 300, { "fit": "contain", "position": "right top", "background": { "r": 255, "g": 150, "b": 0, "alpha": 0.5 }}]]` might be a little intimidating for less-experienced users, but at the same time, I wouldn't really know how to optimize that without reinventing Sharp's full API in our assets layer 🤔  This is just a side effect of exposing the full API of Sharp of course, which in itself is crazy 

PS good thinking on using object-hash for the asset filename. Jealous I didn't think of that earlier! 😄

**Comment by rijkvanzanten** [comment]
Update: I'll add support for the "simple" transforms back in (eg `width`/`height`/`quality`/`format`). These are obviously limited in what they can do (no order, no complex settings), but allow a nice simple API for most transformations. Users who have a need for the more advanced configuration can use the full transforms escape hatch

**Comment by rijkvanzanten** [comment]
Update 2: while the setup using the repeater was nice, it did mean we'd have a total of 3 different syntaxes for providing the transformation options. I've reduced that to a json input (for now, we should add a custom sharp-options interface later) so we have a single API for controlling the transforms

**Review by rijkvanzanten — APPROVED** [review]

**Comment by tim-smart** [comment]
> Update: I'll add support for the "simple" transforms back in (eg `width`/`height`/`quality`/`format`). These are obviously limited in what they can do (no order, no complex settings), but allow a nice simple API for most transformations. Users who have a need for the more advanced configuration can use the full transforms escape hatch

Sound good.

Just looking at the changes, I'm not sure if using the simple params will do anything? Can see where they get resolved :)

**Comment by tim-smart** [comment]
> > Update: I'll add support for the "simple" transforms back in (eg `width`/`height`/`quality`/`format`). These are obviously limited in what they can do (no order, no complex settings), but allow a nice simple API for most transformations. Users who have a need for the more advanced configuration can use the full transforms escape hatch
> 
> Sound good.
> 
> Just looking at the changes, I'm not sure if using the simple params will do anything? Can't see where they get resolved :)

Nevermind, I see what's going on now!

**Pull request merged**